### PR TITLE
cog: add support for "environment" to set environment variables in the generated Dockerfile

### PR DESCRIPTION
**Pull request by hangtwenty**
ℹ️  Docs preview: https://github.com/hangtwenty/cog/blob/cog-environment-variables/docs/yaml.md#environment

[](https://github.com/hangtwenty/cog/blob/cog-environment-variables/docs/yaml.md#environment)

----

### Change 1/2: support for `environment` (`ENV`)

Closes issue #291, "Environment variables in cog.yaml." In Dockerfile parlance, it's support for `ENV`. This supersedes my the previous PR, #361.

This enables adding `environment` to `build` in `cog.yaml`:

```yaml
build:
  environment:
    - EXAMPLE=/src/example
    - DEBUG=1
```

And those become `ENV` directives in the Dockerfile.

### Change 2/2: A single default environment variable to configure caching for PyTorch and some other libraries (`XDG_CACHE_HOME`)

One of the libraries used most often with `cog` is PyTorch. You don't have to do configure anything for it, and now the caching should "just work" in most cases. Likewise for other popular libraries.

<details> <summary>Variables you DON'T need to set </summary>

You don't need to set any of the following because the sane default for `$XDG_CACHE_HOME` will take care of it. (Because it's part of a standard, several libraries use it for their defaults.)

* pytorch: You **don't** need to set `PYTORCH_TRANSFORMERS_CACHE` — [docs](https://pytorch.org/docs/stable/hub.html#:~:text=XDG_CACHE_HOME)
* Hugging Face: You **don't** need to set `TRANSFORMERS_CACHE` — [docs](https://huggingface.co/transformers/v4.0.1/installation.html#caching-models)
    * You _shouldn't_ have to set `PYTORCH_PRETRAINED_BERT_CACHE` — [docs](https://huggingface.co/transformers/v1.0.0/model_doc/overview.html#:~:text=PYTORCH_PRETRAINED_BERT_CACHE)
* pip: no variables needed! Before this PR, `cog` already configured the pip/apt-get cache directories

#### You don't need  to set `TORCH_HOME`

If `TORCH_HOME` is set, that'll take precedence over `XDG_CACHE_HOME` — see [pytorch docs](https://pytorch.org/docs/stable/hub.html#:~:text=XDG_CACHE_HOME), [huggingface docs](https://huggingface.co/transformers/v1.0.0/model_doc/overview.html#:~:text=PYTORCH_PRETRAINED_BERT_CACHE). So, if you set `TORCH_HOME` to something else, just be sure to set it within `/src/`, like `/src/.cache-torch`. That way, it will get cached across runs.

 </details>

### CODE REVIEW

File: pkg/config/config.go
Comment (hangtwenty) [inline pkg/config/config.go:18]: - [ ] Need to remove this note, it was not for this branch.

File: pkg/dockerfile/generator.go
Comment (hangtwenty) [inline pkg/dockerfile/generator.go:142]: - [ ] `TODO(hangtwenty): Clean up`

File: pkg/dockerfile/generator.go
Comment (hangtwenty) [inline pkg/dockerfile/generator.go:179]: - [ ] I should remove the `\` changes here; I think it would cause a layer to rebuild, even for projects that do not adopt `build.env`. Instead, lead the existing `ENV DEBIAN_FRONTEND=...`/`ENV PYTHONUNBUFFERED=`... as separate lines. (In other words, just avoiding unnecessarily invalidating cached layers for models re-built after this change.)

File: pkg/dockerfile/generator.go
Comment (hangtwenty) [inline pkg/dockerfile/generator.go:316]: - [ ] `// TODO(hangtwenty): De-duplicate this - the same function is in config.go`

File: pkg/dockerfile/generator.go
Comment (hangtwenty) [reply pkg/dockerfile/generator.go:142]: On second thought, maybe it's fine? (LMK)

File: pkg/config/config.go
Comment (hangtwenty) [inline pkg/config/config.go]: I removed this comment, not sure why it is showing

File: pkg/config/config.go
Comment (bfirsh) [inline pkg/config/config.go:27]: This should be `environment` for consistency with Docker Compose.

It also probably doesn't need the `Build` prefix because it's already in a struct called `Build`. So maybe `Environment` if we're also renaming to "environment"?

File: pkg/config/config.go
Comment (bfirsh) [inline pkg/config/config.go:210]: Nice!

File: pkg/dockerfile/generator.go
Comment (bfirsh) [inline pkg/dockerfile/generator.go:167]: I wonder if this should be a hard error?

I wonder if we can just defer to Docker for this validation, and just put blindly put in the string into `ENV`?

File: pkg/dockerfile/generator_test.go
Comment (bfirsh) [inline pkg/dockerfile/generator_test.go:180]: I wonder if we could isolate testing the environment stuff with a separate function that is unit tested? If we wanted to change another part of the Dockerfile generation, we now have to update 3 additional tests.

**Review by bfirsh — CHANGES_REQUESTED** [review]
Amazing, thanks @hangtwenty! A few comments inline.

File: pkg/dockerfile/generator_test.go
Comment (hangtwenty) [reply pkg/dockerfile/generator_test.go:180]: Good call, I'll give it a refactor.

File: pkg/dockerfile/generator.go
Comment (hangtwenty) [reply pkg/dockerfile/generator.go:167]: I'm OK with either of those options as well. I'll try them out and compare the experience

File: pkg/config/config.go
Comment (hangtwenty) [reply pkg/config/config.go:27]: <details><summary> For other reviewers /  posterity, here's the discussion about this from the previous PR</summary>

image

(From PR #361 )

</details>

Comment (hangtwenty) [reply pkg/config/config.go:27]: At first I felt differently about this, because `environment` in Docker Compose has its own semantics, but I've come around to agreeing that `build.environment` makes the most sense!

Why I agree now:

* `environment` looks plenty clear in `cog.yaml`; it's under `build`
*  although `env` might resemble `ENV` to me, I think it's because of handwriting lots of Dockerfiles.. but really, `env` is not `ENV`. If someone's experienced with Python (and/or Docker Compose) but not Dockerfiles, `env` might resemble/remind of `virtualenv` or `pyenv`, and that would be confusing.

File: pkg/dockerfile/generator.go
Comment (hangtwenty) [reply pkg/dockerfile/generator.go]: Made it hard-error.

File: pkg/dockerfile/generator_test.go
Comment (hangtwenty) [reply pkg/dockerfile/generator_test.go]: Refactored!

File: pkg/dockerfile/generator_test.go
Comment (hangtwenty) [inline pkg/dockerfile/generator_test.go:87]: ℹ️  This diff is hard to read in "unified" diff, I'd def recommend using side-by-side or viewing the file.

File: pkg/config/config.go
Comment (hangtwenty) [reply pkg/config/config.go]: [It's resolved, but leaving it open for now for easier review.]

Comment (hangtwenty) [reply pkg/config/config.go]: @bfirsh I also revised the Pull Request description so that the usage example is right (`build.environment` not `build.env`)

**Comment by hangtwenty** [comment]
Updated the PR description, so that it can serve as docs for advanced usage.

File: pkg/dockerfile/generator.go
Comment (bfirsh) [inline pkg/dockerfile/generator.go:158]: What's the intention behind this? So it's ordered correctly?

**Review by bfirsh — APPROVED** [review]
Super, thanks @hangtwenty for the fixes! Docs for this would be good too.

File: pkg/dockerfile/generator.go
Comment (hangtwenty) [reply pkg/dockerfile/generator.go:158]: That's right! (Though maybe there's a more idiomatic way to implement it in Go?) 

The order can be significant since environment variables can refer to each other. (As normal with Docker/Dockerfiles.)

**Comment by zeke** [comment]
Nice! It looks like this is ready to go implementation-wise and just needs some documentation.

@hangtwenty is that something you can work on?

**Comment by hangtwenty** [comment]
@zeke Sure, I'll make some time this week.

File: docs/yaml.md
Comment (hangtwenty) [inline docs/yaml.md:30]: "View File" for easier reviewing:

https://github.com/hangtwenty/cog/blob/cog-environment-variables/docs/yaml.md#environment

**Review by zeke — APPROVED** [review]
Love the thorough docs! Thanks for the quick turnaround on that. I left some comments inline.

File: docs/yaml.md
Comment (hangtwenty) [reply docs/yaml.md:41]: Good question — yes order matters. I'll clarify that.

Regarding a realistic example, I'm with you, it's a little tough in this case because I don't want to encourage people to hardcode unnecessary variables, and `XDG_CACHE_HOME` takes care of many realistic use-cases for both PyTorch and Huggingface. I thought about using other realistic examples, but I don't want it to seem like it's endorsing a specific library, since that could get stale.

But I'll try!

Comment (hangtwenty) [reply docs/yaml.md]: OK now it's a more realistic example, while still avoiding encouraging unnecessary vars.

image

File: docs/yaml.md
Comment (hangtwenty) [reply docs/yaml.md]: Great feedback, thank you.

**Comment by hangtwenty** [comment]
image

Oh, I forgot to do the DCO thing. What should I do about that?

**Comment by hangtwenty** [comment]
Hmm, it's getting an ❌  for the DCO check, but I just did `git commit --amend -s` and re-pushed, and it did not resolve it. What do I do about that?

**Comment by zeke** [comment]
> DCO check

I approved it manually. 🍏

**Comment by zeke** [comment]
Great work, @hangtwenty!

I'll defer to @bfirsh and @andreasjansson to make a call on whether this is ready to ship.

**Comment by bfirsh** [comment]
@hangtwenty Thanks for all your help getting this together. I think what is making this hard to review is there are some unrelated changes in this PR that need extra review -- the extra cache environment variables, all the extra docs about how to do caching, etc.

Perhaps we could pull out just the change that adds `environment` to `cog.yaml` and have some really minimal documentation for just that thing, then we could have separate PRs for the other changes.

No worries if you haven't got time -- we want to add this too so maybe we'll have the chance to slim this down.

**Comment by hangtwenty** [comment]
> No worries if you haven't got time -- we want to add this too so maybe we'll have the chance to slim this down.

Feel free, of course! Just confirmed "allow edits by maintainers" is enabled.

**Comment by bfirsh** [comment]
@hangtwenty sorry this got stuck. We'd still love to include this if you ever find the time again for a slimmed down PR. Hope you're well. :)

**Comment by waffletower** [comment]
😢

**Comment by Majdoddin** [comment]
Meanwhile a workaround:
https://github.com/replicate/cog/issues/291#issuecomment-2653025598

**Pull request closed without merging**
